sklearn/metrics/pairwise.py

```diff
@@ -224,41 +234,156 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     """
     X, Y = check_pairwise_arrays(X, Y)
 
-    if X_norm_squared is not None:
-        XX = check_array(X_norm_squared)
-        if XX.shape == (1, X.shape[0]):
-            XX = XX.T
-        elif XX.shape != (X.shape[0], 1):
-            raise ValueError(
-                "Incompatible dimensions for X and X_norm_squared")
-    else:
-        XX = row_norms(X, squared=True)[:, np.newaxis]
+    XX, YY = _check_norms(X, Y, X_norm_squared, Y_norm_squared)
+
+    n_features = X.shape[1]
+
+    # For n_features > 32 we use the 'fast 'method to compute the euclidean
+    # distance, i.e. d(x,y)² = ||x||² + ||y||² - 2 * x.y
+    # It's faster but less precise.
+    if n_features > 32:
 
-    if X is Y:  # shortcut in the common case euclidean_distances(X, X)
-        YY = XX.T
-    elif Y_norm_squared is not None:
-        YY = np.atleast_2d(Y_norm_squared)
+        # To minimize precision issues with float32, we compute the distance
+        # matrix on chunks of X and Y upcasted to float64
+        if X.dtype == np.float32:
+            distances = _euclidean_distances_upcast_fast(X, XX, Y, YY)
 
-        if YY.shape != (1, Y.shape[0]):
-            raise ValueError(
-                "Incompatible dimensions for Y and Y_norm_squared")
+        # if dtype is already float64, no need to chunk and upcast
+        else:
+            if X is Y and not issparse(X):
+                # In this case the distance matrix is symmetric, so we only
+                # need to compute half of it. When X is dense, we can benefit
+                # from the BLAS triangular matrix matrix multiplication `syrk`.
+                distances = _euclidean_dense_dense_fast_sym(X, XX)
+            else:
+                distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)
+                _add_norms(distances, XX, YY)
```

Comment:
It's not much better, but it's a little bit better because of the multiple passes over the data and the broadcasting in
```
distances += XX[:, np.newaxis]
distances += YY[np.newaxis, :]
np.maximum(distances, 0, out=distances)
```
If the 10% speedup is not worth I can revert

Reply:
It's not a 10% speed up on the total distance calculation? I'd be inclined towards readability in this case.

Reply:
Well it can be on the total time actually. For example
```python
X = np.random.RandomState(0).random_sample((100000, 64)) 
Y = np.random.RandomState(1).random_sample((100, 64)) 
%timeit euclidean_distances(X,Y)
```
adding norms in cython: 88.7 ms ± 180 µs
adding norms with numpy: 105 ms ± 1.42 ms

But when the number of features increases, the matrix matrix multiplication takes a more important part and the difference decreases. I'll keep the readable code for now. But it might be interesting for a later PR :)